blockstack_client/backend/drivers/sia.py

```diff
@@ -0,0 +1,524 @@
+    """
+
+    data_id = urllib.quote(data_id.replace('/', '-2f'))  # Hex code for forward slash.
+
+    return "http://sia.tech/%s" % data_id
+
+
+def get_immutable_handler(data_hash, **kw):
+    """
+    Given a cryptographic hash of some data, go and fetch it.
+    This is used by the immutable data API, whereby users can 
+    add and remove data hashes in their zone file (hence the term
+    "immutable").  The method that puts data for this method
+    to fetch is put_immutable_handler(), described below.
+ 
+    Drivers are encouraged but not required to implement this method.
+    A common strategy is to treat the data_hash like the data_id in
+    make_mutable_url().
+ 
+    **kw contains hints from Blockstack about the nature of the request.
+    TODO: document them here.
+ 
+    Returns the data on success.  It must hash to data_hash (sha256)
+    Returns None on error.  Does not raise an exception.
+    """
+
+    return download("immutable-%s" % data_hash)
+
+
+def get_mutable_handler(uri, **kw):
+    """
+    Given the URL to some data generated by an earlier call to
+    make_mutable_url().
+ 
+    **kw contains hints from Blockstack about the nature of the request.
+    TODO: document them here.
+ 
+    Drivers are encouraged but not required to implement this method.
+ 
+    Returns the data on success.  The driver is not expected to e.g. verify
+    its authenticity (Blockstack will take care of this).
+    Return None on error.  Does not raise an exception.
+    """
+
+    key = urlparse.urlparse(uri).path[1:]
+    return download(key)
+
+
+def upload(key, data, txid):
+    global SIAD_HOST, SIAD_PORT, USER_AGENT, SIAD_PASSWD
+
+    key = urllib.quote(key.replace("/", r"-2f"))
+
+    sia_upload = "http://%s:%s/renter/upload/%s" % (SIAD_HOST, SIAD_PORT, key)
+
+    import tempfile
+    with tempfile.NamedTemporaryFile() as temp:
+        log.debug("[%s] Preparing upload to siad @ %s..." % (txid, sia_upload))
+
+        temp.write(data)
+        temp.flush()
+
+        ok = False
+
+        try:
+            r = requests.post(sia_upload, params={
+                'source': temp.name
+            }, headers={
+                'user-agent': USER_AGENT
+            }, auth=('', SIAD_PASSWD))
+
+            log.debug("[%s] Uploading to siad @ %s..." % (txid, r.url))
+
+            ok = r.status_code == requests.codes.no_content
+
+            if not ok:
+                log.debug("failed to upload file to siad. Status: %s - Response: %s", r.status_code, r.json())
+        except Exception as e:
+            log.exception(e)
+
+        return ok
+
+
+def download(key):
+    global SIAD_HOST, SIAD_PORT, USER_AGENT, SIAD_PASSWD
+
+    key = urllib.quote(key.replace("/", r"-2f"))
+
+    sia_download = "http://%s:%s/renter/download/%s" % (SIAD_HOST, SIAD_PORT, key)
+
+    log.debug("Preparing download from siad @ %s..." % sia_download)
+
+    from os.path import expanduser
+    destination = expanduser("~") + "/%s" % key
```

Comment:
I started down the temp file path first but the way that temp files work in Python is that they're created with permissions such that only the files creator can write to it and since the siad daemon could be running as a different user it'd be problematic to do it that way.